### DISCUSSION TITLE
Topics to check (meet 08/09)

### DISCUSSION THREAD

**Discussion by wellington36** [wellington36/tweedie-numerical-approximation#8] [General]
### 2. Check the origin of the tweedie pdf series form

$$
\sum_{k = 1}^\infty \frac{\Gamma(1 + \alpha k)}{\Gamma(1 + k)} (x(-1/z, \alpha))^k \sin(-k \pi \alpha);
$$

### 5 (extra). Some possible applications of tweedie to check

- https://www.tandfonline.com/doi/suppl/10.1198/jasa.2011.tm11181
- https://www.tandfonline.com/doi/full/10.1080/03610926.2021.1948061

# Referencies
  [1]<a id="dias"></a>: DIAS NL, RIBEIRO JR PJ. Practical rules for summing the series of the Tweedie probability density function with high-precision arithmetic. An Acad Bras Ciênc [Internet]. 2019;91(4):e20180268. Available from: https://doi.org/10.1590/0001-3765201920180268
  
  [2]<a id="dunn2005"></a>: Dunn, P. K. and Smyth, G. K. (2005). Series evaluation of tweedie exponential dispersion model densities. Statistics and Computing, 15(4):267–280.

  [3]<a id="dunn2008"></a>: Dunn, P. K. and Smyth, G. K. (2008). Evaluation of tweedie exponential dispersion model
densities by fourier inversion. Statistics and Computing, 18(1):73–86.

**Comment by wellington36** [comment]
> ### 1. Evaluate the tweedie pdf series in the ratio test;

In [[1]](#dias) the authors present a proof that the series (in the form presented in it) is absolutely convergent by the ratio test.

# Referencies
  [1]<a id="dias"></a>: DIAS NL, RIBEIRO JR PJ. Practical rules for summing the series of the Tweedie probability density function with high-precision arithmetic. An Acad Bras Ciênc [Internet]. 2019;91(4):e20180268. Available from: https://doi.org/10.1590/0001-3765201920180268

**Comment by maxbiostat** [comment]
"In [[1]](https://github.com/wellington36/tweedie-numerical-approximation/discussions/8#dias) the authors present a proof that the series (in the form presented in it) is absolutely convergent by the ratio test."

**Comment by wellington36** [comment]
> ### 2. Check the origin of the tweedie pdf series form
> $$
> \sum_{k = 1}^\infty \frac{\Gamma(1 + \alpha k)}{\Gamma(1 + k)} (x(-1/z, \alpha))^k \sin(-k \pi \alpha);
> $$

This way of writing the tweedie's pdf comes from [[4]](#jorg1992) using Fourier inverse (this explains the existence of the sine function). In the traditional way as integral:

![image](https://github.com/wellington36/tweedie-numerical-approximation/assets/61877847/fd9a3cff-d885-499d-84f7-ad52dffa84ab)
_from: https://en.wikipedia.org/wiki/Tweedie_distribution_

or as presented in  [[2]](#dunn2005)

$$
f(y; \theta, \sigma) = a(y, \sigma) exp \left [ \frac{1}{\sigma} \{ y \theta - \kappa(\theta) \} \right],
$$

for given functions $\kappa$ and $a()$ [[5]](#jorg1997). This version in series format is used in cases like [[1]](#days), [[2]](#dunn2005) and [[3]](#dunn2008) quoting Jørgensen.

# Referencies
  [1]<a id="dias"></a>: DIAS NL, RIBEIRO JR PJ. Practical rules for summing the series of the Tweedie probability density function with high-precision arithmetic. An Acad Bras Ciênc [Internet]. 2019;91(4):e20180268. Available from: https://doi.org/10.1590/0001-3765201920180268
  
  [2]<a id="dunn2005"></a>: Dunn, P. K. and Smyth, G. K. (2005). Series evaluation of tweedie exponential dispersion model densities. Statistics and Computing, 15(4):267–280.

  [3]<a id="dunn2008"></a>: Dunn, P. K. and Smyth, G. K. (2008). Evaluation of tweedie exponential dispersion model
densities by fourier inversion. Statistics and Computing, 18(1):73–86.

  [4]<a id="jorg1992"></a>: JØRGENSEN B. 1992. The theory of exponential dispersion models and analysis of deviance. 2nd edition. Rio de Janeiro: Instituto de Matemática Pura e Aplicada.

  [5]<a id="jorg1997"></a>: Jørgensen B. 1997. The Theory of Dispersion Models, Chapman and Hall, London.

**Comment by wellington36** [comment]
> ### 3. Check if the quadrature done in [[1]](#dias) is analogous to that of [scypi.integrate.quad](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.quad.html).

In the `quad` documentation, the `points` argument function is not the one we need to pass the number of points to be used in the Gauss-Legendre method:

> **points: (sequence of floats,ints), optional**
> A sequence of break points in the bounded integration interval where local difficulties of the integrand may occur (e.g., singularities, discontinuities). The sequence does not have to be sorted. Note that this option cannot be used in conjunction with weight.

To do as mentioned in the paper, we can use the function [scipy.special.roots_legendre](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.roots_legendre.html) ("Compute the sample points and weights for Gauss-Legendre quadrature"). Hence, as our pdf function receives only one parameter at a time, we can do the integration as in the paper as follows:

```python
from scipy.special import roots_legendre
def gauss_legendre_integration(f, a, b, n_points):
    x, w = roots_legendre(n_points)
    integral = 0.0
    
    for i in range(n_points):
        integral += w[i] * f(0.5 * (b - a) * x[i] + 0.5 * (b + a))
    integral *= 0.5 * (b - a)
    
    return integral
```

This will be solved in https://github.com/wellington36/tweedie-numerical-approximation/issues/9.

**Comment by wellington36** [comment]
> ### 4. What's the stopping criteria used in [[2]](#dunn2005) and [[3]](#dunn2008).
## Dias's summation
This topic is a little more delicate than the previous ones. First, we recall the stopping criterion of [[1]](#dias), given the series

$$
\sum_{k=1}^\infty \frac{\Gamma(1 + \alpha k)}{\Gamma(1 + k)} (x(-1/z, \alpha))^k \sin(-k \pi \alpha)
$$

where,

$$
x(\theta, \alpha) = \frac{\alpha - 1}{\alpha} \left( \frac{\theta}{\alpha - 1} \right)^\alpha.
$$

Here $\alpha = \frac{p - 2}{p - 1}$, and the stopping criterion summing from 1 to k, is if

$$
\frac{b_k}{|S(k)|} < 10^{-10}
$$

where,

$$
b_k =  \frac{\Gamma(1 + \alpha k)}{\Gamma(1 + k)} (|x(-1/z, \alpha)|)^k,
$$

and $S(k)$ is the partial sum of the series to k.

## Dunn's summation


[[2]](#dunn2005) and [[3]](#dunn2008) we have 2 series, one for each interval of p:

- **(1)** for $1 < p < 2$: $W(z, \sigma, p) = {\sum}^{\infty}_{j=1}W_j$, where

$$
W_j = \frac{z^{j \alpha} (p-1)^{\alpha j}}{\sigma^{j (1 \alpha)} (2 - p)^j j! \Gamma(-j \alpha)}.
$$

- **(2)** for $p > 2$: $V(z, \sigma, p) = {\sum}^{\infty}_{k=1} V_k$, where

$$
V_k = \frac{\Gamma(1 + \alpha k) \sigma^{k (\alpha  - 1)} (p - 1)^{\alpha k}}{\Gamma(k) (p - 2)^k z^{\alpha k}} (-1)^k \sin (-k \pi \alpha)
$$

The idea is instead of adding starting at 1 up to a fixed number, is to find the maximum and start adding around it. Thus, we avoid adding unnecessary terms. For the case **(1)** the function that describes the series is continuous, then we can derivate and find de maximum value. For the case **(2)** (the target in this  repo), this function isn't continuous, but if we define

$$
V_{env}(k) = \frac{w^k \Gamma(1 + \alpha k)}{\Gamma(1 + k)},
$$

where,

$$
w = \frac{(p-1)^\alpha \sigma^{\alpha - 1}}{z^\alpha (p-1)},
$$

we are just removing $(-1)^k$ and $\sin$. The definition ensures that $|V_k| \leq V_{env}(k), \forall k$. Using also the [Stiling's approximation](https://en.wikipedia.org/wiki/Stirling%27s_approximation), we can find de maximum

$$
k_{max} = \frac{z^{2 - p}}{\sigma (p-2)}
$$

and

$$
\log V_{max} = (1 - \alpha) k_{max} + \frac{1}{2} \log\alpha
$$

## Dunn's stopping criteria

Following [[2]](#2005dunn) The objective is evaluating $W(z, \sigma, p)$ with

$$
\widetilde{W}(z, \sigma, p)  = \sum_{j=j_L}^{j_U} W_j
$$

and $V(z, \sigma, p)$ with

$$
\widetilde{V}(z, \sigma, p)  = \sum_{k=k_L}^{k_U} V_j
$$

where $j_L, j_U, k_L, k_U$ are chosen limits. For the first case, $\frac{\partial \log W_j}{\partial j}$ is monotonic decreasing implies that $\log W_j$ is strictly convex as a function of j and hence that the $W_j$ decay faster than geometrically on either side of $j_{max}$. The approximation error can therefore be bounded by geometric sums,

$$
W(z, \sigma, p) - \widetilde{W}(z, \sigma, p) < W_{j_L-1} \frac{1 - r_L^{j_L-1}}{1 - r_L} + W_{j_U-1} \frac{1}{1-r_U}
$$

where

$$
r_L = exp \left . \left(  \frac{\partial \log W_j}{\partial j} \right) \right |_{j = j_L - 1}
$$

and

$$
r_U = exp \left . \left(  \frac{\partial \log W_j}{\partial j} \right) \right |_{j = j_U - 1}
$$

From [[2]](#2005dunn) the same bound can be constricted for $V(z, \sigma, p) - \widetilde{V}(z, \sigma, p)$ in terms of $V_{env}(k)$. In practice, we can choose this bounds with $j_L < j_{max} < j_U$ such that $W_{j_L}, W_{j_U} < \varepsilon W_{max}$ and for the second case $k_L < k_{max} < k_U$ such that $W_{j_L}, W_{j_U} < \varepsilon W_{max}$. Here $\varepsilon = 10^{-16}$ would ensure double precision accuracy in 64-bit floating pont. In practice we use $\varepsilon = exp(-37) \approx 8 \times 10^{-17}$, using log-scale.
 
# Referencies
  [1]<a id="dias"></a>: DIAS NL, RIBEIRO JR PJ. Practical rules for summing the series of the Tweedie probability density function with high-precision arithmetic. An Acad Bras Ciênc [Internet]. 2019;91(4):e20180268. Available from: https://doi.org/10.1590/0001-3765201920180268
  
  [2]<a id="dunn2005"></a>: Dunn, P. K. and Smyth, G. K. (2005). Series evaluation of tweedie exponential dispersion model densities. Statistics and Computing, 15(4):267–280.

  [3]<a id="dunn2008"></a>: Dunn, P. K. and Smyth, G. K. (2008). Evaluation of tweedie exponential dispersion model
densities by fourier inversion. Statistics and Computing, 18(1):73–86.
